mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+        )
+        q_loss_info["mean_first_reward"] = jnp.mean(first_reward)
+        q_loss_info["mean_next_qval"] = jnp.mean(next_q_val)
+        q_loss_info["done"] = jnp.mean(data.term_or_trunc)
+
+        # Mean over the device and batch dimension.
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="device")
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="batch")
+        q_updates, next_opt_state = opt.update(q_grads, opt_states)
+        (next_online_params, next_mixer_params) = optax.apply_updates(
+            (params.online, params.mixer_online), q_updates
+        )
+
+        # Target network update.
+        if cfg.system.hard_update:
+            next_target_params = optax.periodic_update(
+                next_online_params, params.target, t_train, cfg.system.update_period
+            )
+            next_mixer_target_params = optax.periodic_update(
+                next_mixer_params, params.mixer_target, t_train, cfg.system.update_period
+            )
+        else:
+            next_target_params = optax.incremental_update(
+                next_online_params, params.target, cfg.system.tau
+            )
+            next_mixer_target_params = optax.incremental_update(
+                next_mixer_params, params.mixer_target, cfg.system.tau
+            )
+        # Repack params and opt_states.
+        next_params = QMIXParams(
+            next_online_params,
+            next_target_params,
+            next_mixer_params,
+            next_mixer_target_params,
+        )
+
+        return next_params, next_opt_state, q_loss_info
+
+    def train(
+        train_state: TrainState[QMIXParams], _: Any
+    ) -> Tuple[TrainState[QMIXParams], Metrics]:
+        """Sample, train and repack."""
+
+        buffer_state, params, opt_states, t_train, key = train_state
+        next_key, buff_key = jax.random.split(key, 2)
+
+        data = rb.sample(buffer_state, buff_key).experience
+
+        # Learn
+        next_params, next_opt_states, q_loss_info = update_q(params, opt_states, data, t_train)
+
+        next_train_state = TrainState(
+            buffer_state, next_params, next_opt_states, t_train + 1, next_key
+        )
+
+        return next_train_state, q_loss_info
+
+    # ---- Act-train loop ----
+    scanned_act = lambda state: lax.scan(action_step, state, None, length=cfg.system.rollout_length)
+    scanned_train = lambda state: lax.scan(train, state, None, length=cfg.system.epochs)
+
+    # Act and train
+    def update_step(
+        learner_state: LearnerState, _: Any
+    ) -> Tuple[LearnerState, Tuple[Metrics, Metrics]]:
+        """Act, then learn."""
+
+        (
+            obs,
+            terminal,
+            term_or_trunc,
+            hidden_state,
+            env_state,
+            time_steps,
+            train_steps,
+            opt_state,
+            buffer_state,
+            params,
+            key,
+        ) = learner_state
```

Comment:
Half of this function is packing and unpacking params :joy:

Reply:
It's true :sweat_smile: